### PR TITLE
Add optional count argument to `Semaphore::post`

### PR DESCRIPTION
**Pull request by radiantgurl**
Implements a new optional argument in Semaphore.post to allow resuming multiple threads at once.

### CODE REVIEW

File: core/core_bind.cpp [hunk 1]
```diff
@@ -1209,14 +1212,15 @@ bool Semaphore::try_wait() {
 	return semaphore.try_wait();
 }
 
-void Semaphore::post() {
-	semaphore.post();
+void Semaphore::post(int amount) {
```
Comment (RandomShaper) [inline core/core_bind.cpp:1215]:
```suggestion
void Semaphore::post(int p_count) {
```

File: core/core_bind.cpp [hunk 2]
```diff
@@ -1209,14 +1212,15 @@ bool Semaphore::try_wait() {
...
+	ERR_FAIL_COND(amount > 0);
```
Comment (RandomShaper) [inline core/core_bind.cpp:1216]:
```suggestion
	ERR_FAIL_COND(p_count <= 0);
```

**Comment by Mickeon** [comment]
May I ask in what occasion would this be useful? It is a relatively innocent change, but it's still worth knowing _why_ it would need to be done.

**Comment by RadiantUwU** [comment]
It's just so you dont have to do a for loop anymore to resume more and more threads, and it would be way more optimized than constantly locking and unlocking the mutex, adding 1 to the count and running `notify_one`.

File: core/core_bind.cpp (hunk 2 above)
Comment (RadiantUwU) [reply core/core_bind.cpp:1216]: Woopsies

File: core/core_bind.compat.inc [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+void Semaphore::_post_bind_compat_93605() {
+	semaphore.post();
+}
+
+void Semaphore::_bind_compatibility_methods() {
+	ClassDB::bind_compatibility_method(D_METHOD("post"), &Semaphore::_post_bind_compat_93605);
+}
+
+}; // namespace core_bind
```
Comment (AThousandShips) [inline core/core_bind.compat.inc:43]:
```suggestion
}; // namespace core_bind

#endif
```

File: core/core_bind.cpp [hunk 4]
```diff
@@ -30,6 +30,9 @@
 
 #include "core_bind.h"
 
+#include "core/object/object.h"
+#include "core_bind.compat.inc"
```
Comment (AThousandShips) [inline core/core_bind.cpp:34]:
```suggestion
```

File: core/core_bind.cpp [hunk 5]
```diff
@@ -30,6 +30,9 @@
 
 #include "core_bind.h"
```
Comment (AThousandShips) [inline core/core_bind.cpp:31]:
```suggestion
#include "core_bind.h"
#include "core_bind.compat.inc"
```

File: core/core_bind.cpp [hunk 6]
```diff
@@ -30,6 +30,9 @@
...
 
+#include "core/object/object.h"
```
Comment (AThousandShips) [inline core/core_bind.cpp:33]:
```suggestion
```
Already included 

Are you using some tool in your IDE to automatically add includes? I'd suggest turning it off or checking that it actually makes the right suggestions, this is already included

File: core/core_bind.h [hunk 7]
```diff
@@ -389,12 +389,17 @@ class Semaphore : public RefCounted {
 	GDCLASS(Semaphore, RefCounted);
 	::Semaphore semaphore;
 
+protected:
 	static void _bind_methods();
+	static void _bind_compatibility_methods();
+
+private:
+	void _post_bind_compat_93605();
```
Comment (AThousandShips) [inline core/core_bind.h:397]:
```suggestion
```

File: core/core_bind.h [hunk 8]
```diff
@@ -389,12 +389,17 @@ class Semaphore : public RefCounted {
 	GDCLASS(Semaphore, RefCounted);
 	::Semaphore semaphore;
 
+protected:
 	static void _bind_methods();
+	static void _bind_compatibility_methods();
```
Comment (AThousandShips) [inline core/core_bind.h:394]:
```suggestion
#ifndef DISABLE_DEPRECATED
	void _post_bind_compat_93605();
	static void _bind_compatibility_methods();
#endif // DISABLE_DEPRECATED
```

**Review by RandomShaper — APPROVED** [review]
Leaving my approval on the added API. This needs to have the compatibility stuff fixed (which I'm not very familiar with) fixed per @AThousandShips comments before merging.

**Comment by Bromeon** [comment]
Was a named method considered (`post_n` or something)? Not sure what's the usual API design in such cases.

File: core/core_bind.h [hunk 9]
```diff
@@ -389,12 +389,18 @@ class Semaphore : public RefCounted {
...
 	::Semaphore semaphore;
 
+protected:
 	static void _bind_methods();
+#ifndef DISABLE_DEPRECATED
+	void _post_bind_compat_93605();
+	static void _bind_compatibility_methods();
+#endif // DISABLE_DEPRECATED
 
+private:
```
Comment (AThousandShips) [inline core/core_bind.h:399]:
```suggestion
```
Don't need this

File: misc/extension_api_validation/4.2-stable.expected [hunk 10]
```diff
@@ -372,3 +372,10 @@ GH-91382
 Validate extension JSON: Error: Field 'classes/AudioStreamPlaybackPolyphonic/methods/play_stream/arguments': size changed value in new API, from 4 to 6.
 
 Optional arguments added. Compatibility methods registered.
+
+
+GH-93605
+--------
+Validate extension JSON: JSON file: Field was added in a way that breaks compatibility 'classes/Semaphore/methods/post': arguments
```
Comment (AThousandShips) [inline misc/extension_api_validation/4.2-stable.expected:379]: This isn't the line to add, I think it should be:
```suggestion
Validate extension JSON: Error: Field 'classes/Semaphore/methods/post/arguments': size changed value in new API, from 0 to 1.
```
But we can see when the CI is done

Comment (AThousandShips) [reply misc/extension_api_validation/4.2-stable.expected:379]: My bad was the correct one, just hadnt seen before

**Review by AThousandShips — APPROVED** [review]

File: doc/classes/Semaphore.xml [hunk 11]
```diff
@@ -17,8 +17,9 @@
 	<methods>
 		<method name="post">
 			<return type="void" />
+			<param index="0" name="count" type="int" default="1" />
 			<description>
-				Lowers the [Semaphore], allowing one more thread in.
+				Lowers the [Semaphore], allowing one or more thread(s) in.
```
Comment (Mickeon) [inline doc/classes/Semaphore.xml:22]: No need for those parentheses I feel.
```suggestion
				Lowers the [Semaphore], allowing one or more threads in.
```
Alternatively,
```suggestion
				Lowers the [Semaphore], allowing one thread in, or more if [param count] is specified.
```

**Review by Mickeon — COMMENTED** [review]
Is `count` as the name of the parameter completely fine? Yyyyeah, I'm sure it'll be fine.

File: doc/classes/Semaphore.xml [hunk 12]
```diff
@@ -17,8 +17,9 @@
 	<methods>
 		<method name="post">
 			<return type="void" />
+			<param index="0" name="count" type="int" default="1" />
 			<description>
-				Lowers the [Semaphore], allowing one more thread in.
+				Lowers the [Semaphore], allowing one thread in, or more if [param count] is specified
```
Comment (AThousandShips) [inline doc/classes/Semaphore.xml:22]:
```suggestion
				Lowers the [Semaphore], allowing one thread in, or more if [param count] is specified.
```

**Comment by AThousandShips** [comment]
The co-author additions for just reviews aren't necessary by the way, feel free to remove them

**Comment by Mickeon** [comment]
I think we can safely push this to 4.4 . Still waiting on the GDExtension and DotNet teams for review.

**Comment by akien-mga** [comment]
Thanks!

**Comment by RandomShaper** [comment]
I've found something a bit concerning... One of my PRs was tested against 4.3, where it worked. On _master_ it broke. The reason was that on _master_ I had to add `.bind(1)` to a `Callable` to `core_bind::Semaphore::post()`; otherwise, I would get too-few-args error. On 4.3, adding the `.bind(1)` makes the code fail with too-many-args error. Therefore, there's a compat breakage here. Or is the reason that we have an issue with the binding-call system not honoring the default arg in case the `.bind(1)` is missing?

**Comment by AThousandShips** [comment]
It's been discussed (forget where) and I've looked at some tricks to make it easier by creating a custom callable from a `MethodBind` but haven't done much on it as it's a bit convoluted, and would require fetching the method (also wouldn't work with non-bound methods, though we could use the bind without using the ClassDB system)

But will see about opening some issue to track it and look at ideas when I can

**Comment by Mickeon** [comment]
If this is a severe problem reverting this PR should be fine as it's a relatively small Quality-of-Life feature.

**Pull request merged**